File: Tron/Tron/Client/Cycles/CycleMovementController.ts
/// <reference path="../Collidables/MovementControllers/MovementController.ts" />
/// <reference path="../Space/Map.ts" />
/// <reference path="../Extensions/Number.ts" />

class CycleMovementController extends MovementController {
    static MAX_SPEED: number = 1000;
    static Y_OFFSET: number = 35;

    static HALF_PI: number = Math.PI / 2;
    static TWO_PI: number = Math.PI * 2;

    static Velocities: IVector3[];

    public HeadLocation: MapLocation;
    public Context: IMesh;

    constructor(private _context: IMesh, startVelocity: IVector3) {
        super(_context, startVelocity, CycleMovementController.MAX_SPEED);

        this.Context = this._context;
        this._context.position.y = 50;//CycleMovementController.Y_OFFSET;

        if (!CycleMovementController.Velocities) {
            this.calculateVelocities();
        }
    }

    private calculateVelocities(): void {
        CycleMovementController.Velocities = [];
        CycleMovementController.Velocities[0] = new THREE.Vector3(0, 0, -CycleMovementController.MAX_SPEED);
        CycleMovementController.Velocities[Math.round(CycleMovementController.HALF_PI)] = new THREE.Vector3(-CycleMovementController.MAX_SPEED, 0, 0);
        CycleMovementController.Velocities[Math.round(Math.PI)] = new THREE.Vector3(0, 0, CycleMovementController.MAX_SPEED);
        CycleMovementController.Velocities[Math.round(Math.PI + CycleMovementController.HALF_PI)] = new THREE.Vector3(CycleMovementController.MAX_SPEED, 0, 0);
    }

    private stabilizeValue(position: any, velocity: any): number {
        if (velocity.normalized() * position.normalized() > 0)
        {
            position -= position % Map.FLOOR_TILE_SIZE.Width;
        }
        else
        {
            if (position != 0)
            {
                position -= position % Map.FLOOR_TILE_SIZE.Width - Map.FLOOR_TILE_SIZE.Width * position.normalized();
            }
        }

        return position;
    }

    private positionOnLine(): void {
        this._context.position = this.GetLinePosition(this._context.position);
    }

    public GetLinePosition(currentPosition: IVector3): IVector3 {
        var currentVelocity: IVector3;

        currentPosition = currentPosition.clone();

        // If our velocity was zero then deterine the velocity based on the current rotation (This happens when we've collided)
        if (this.Velocity.isZero())
        {
            return Map.Utilities.ToPosition(this.HeadLocation, currentPosition.y);
        }
        else
        {
            currentVelocity = this.Velocity;
        }

        if (currentVelocity.z != 0)
        {
            currentPosition.z = this.stabilizeValue(currentPosition.z, currentVelocity.z);
        }
        else if (currentVelocity.x != 0)
        {
            currentPosition.x = this.stabilizeValue(currentPosition.x, currentVelocity.x);
        }

        return currentPosition;
    }

    public Move(direction: string): void {
        this.positionOnLine();

        if (direction === "Left") {
            this._context.rotation.y = (this._context.rotation.y + CycleMovementController.HALF_PI) % CycleMovementController.TWO_PI;

            if (Math.round(this._context.rotation.y) == 6) { // Above two pi            
                this._context.rotation.y = 0;
            }
        }
        else if (direction === "Right") {
            this._context.rotation.y -= CycleMovementController.HALF_PI;

            if (this._context.rotation.y < 0) {
                if (Math.round(this._context.rotation.y) === 0) {
                    this._context.rotation.y = 0;
                }
                else {
                    this._context.rotation.y += CycleMovementController.TWO_PI;
                }
            }
        }

        this.Velocity = CycleMovementController.Velocities[Math.round(this._context.rotation.y)];
    }

    public Update(gameTime: GameTime): void {
        super.Update(gameTime);

        var incrementor: IVector3 = this.Velocity.clone().multiplyScalar(gameTime.FractionOfSecond);
        this._context.position.addSelf(incrementor);
    }
}

File: Tron/TronGameServer/Cycles/CycleMovementController.cs
using System;
using System.Collections.Generic;
using Tron.Utilities;

namespace Tron.GameServer
{
    public class CycleMovementController : MovementController, IDisposable
    {
        private GameConfiguration _gameConfiguration;
        private Dictionary<double, Vector3> _velocities;
        private Map _map;

        public CycleMovementController(Vector3 position, Vector3 velocity, double rotation, Map map, GameConfiguration gameConfiguration)
            : base(position, velocity, Math.Round(rotation, 4))
        {
            _gameConfiguration = gameConfiguration;
            _map = map;
            // There aren't many velocities we can have so calculate them prior to cycle activation
            calculateVelocities();
        }

        public Vector3 RequestedPosition { get; set; }
        public MapLocation HeadLocation { get; set; }

        private void calculateVelocities()
        {
            _velocities = new Dictionary<double, Vector3>();
            _velocities.Add(0, new Vector3(0, 0, -_gameConfiguration.CycleConfig.MAX_SPEED));
            _velocities.Add(Math.Round(TMath.HALF_PI), new Vector3(-_gameConfiguration.CycleConfig.MAX_SPEED, 0, 0));
            _velocities.Add(Math.Round(TMath.PI), new Vector3(0, 0, _gameConfiguration.CycleConfig.MAX_SPEED));
            _velocities.Add(Math.Round(TMath.ONE_AND_ONE_HALF_PI), new Vector3(_gameConfiguration.CycleConfig.MAX_SPEED, 0, 0));
        }

        private double stabilizeValue(double position, double velocity)
        {
            if (velocity.Normalized() * position.Normalized() > 0)
            {
                position -= position % _gameConfiguration.MapConfig.FLOOR_TILE_SIZE.Width;
            }
            else if(position != 0)
            {
                position -= position % _gameConfiguration.MapConfig.FLOOR_TILE_SIZE.Width - _gameConfiguration.MapConfig.FLOOR_TILE_SIZE.Width * position.Normalized();
            }

            return position;
        }

        private void positionOnLine()
        {
            Position = GetLinePosition(Position);
        }

        public void ConfirmPositionRequest(Vector3 alteredPosition = null)
        {
            Position = alteredPosition ?? RequestedPosition;
            RequestedPosition = null;
        }

        public Vector3 GetLinePosition(Vector3 currentPosition)
        {
            Vector3 currentVelocity;
            currentPosition = currentPosition.Clone();

            // If our velocity was zero then reset position to where our head location is
            if (Velocity.IsZero())
            {
                return _map.Utilities.ToPosition(HeadLocation, currentPosition.y);
            }
            else
            {
                currentVelocity = Velocity;
            }
            
            if (currentVelocity.z != 0)
            {
                currentPosition.z = stabilizeValue(currentPosition.z, currentVelocity.z);
            }
            else if (currentVelocity.x != 0)
            {
                currentPosition.x = stabilizeValue(currentPosition.x, currentVelocity.x);
            }

            return currentPosition;
        }

        public bool CanMove(MovementFlag direction)
        {
            return GetValidMovements().Contains(direction);
        }

        public List<MovementFlag> GetValidMovements()
        {
            var validMovements = new List<MovementFlag>();
            var rotation = Math.Round(Rotation);
            var locationCheck = HeadLocation.Clone();

            if (rotation == 0) // Going up
            {
                locationCheck.Column--;
                if (_map.Empty(locationCheck))
                {
                    validMovements.Add(MovementFlag.Left);
                }

                locationCheck.Column += 2;
                if (_map.Empty(locationCheck))
                {
                    validMovements.Add(MovementFlag.Right);
                }
            }
            else if (rotation == 2) // Going Left
            {
                locationCheck.Row++;
                if (_map.Empty(locationCheck))
                {
                    validMovements.Add(MovementFlag.Left);
                }

                locationCheck.Row -= 2;
                if (_map.Empty(locationCheck))
                {
                    validMovements.Add(MovementFlag.Right);
                }
            }
            else if (rotation == 3) // Going Down
            {
                locationCheck.Column++;
                if (_map.Empty(locationCheck))
                {
                    validMovements.Add(MovementFlag.Left);
                }

                locationCheck.Column -= 2;
                if (_map.Empty(locationCheck))
                {
                    validMovements.Add(MovementFlag.Right);
                }
            }
            else if (rotation == 5) // Going Right
            {
                locationCheck.Row--;
                if (_map.Empty(locationCheck))
                {
                    validMovements.Add(MovementFlag.Left);
                }

                locationCheck.Row += 2;
                if (_map.Empty(locationCheck))
                {
                    validMovements.Add(MovementFlag.Right);
                }
            }

            return validMovements;
        }

        public void Move(MovementFlag direction)
        {
            positionOnLine();

            if (direction == MovementFlag.Left)
            {
                Rotation = (Rotation + TMath.HALF_PI) % TMath.TWO_PI;

                if(Math.Round(Rotation) == 6) // Above two pi
                {
                    Rotation = 0;
                }
            }
            else if (direction == MovementFlag.Right)
            {
                Rotation -= TMath.HALF_PI;

                if (Rotation < 0)
                {
                    if (Math.Round(Rotation) == 0) 
                    {
                        Rotation = 0;
                    }
                    else
                    {
                        Rotation += TMath.TWO_PI;
                    }
                }
            }

            Velocity = _velocities[Math.Round(Rotation)];           
        }

        public override void Update(GameTime gameTime)
        {
            base.Update(gameTime);

            var incrementor = Velocity * gameTime.FractionOfSecond;
            RequestedPosition = Position + incrementor;
        }

        public override void Dispose()
        {
            _map = null;
            _gameConfiguration = null;
            _velocities.Clear();
            _velocities = null;
            base.Dispose();
        }
    }
}
